src/boot/efi/efi-string.c

```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+
+        bool alternative_form = false, have_field_width = false, long_arg = false;
+        size_t field_length = sizeof(int);
+        int precision = -1, field_width = -1;
+
+        for (;; format++)
+                switch (*format) {
+                case '0':
+                        *ret_pad_zero = true;
+                        break;
+                case '.':
+                        have_field_width = true;
+                        break;
+                case '#':
+                        alternative_form = true;
+                        break;
+
+                case '*':
+                case '1' ... '9': {
+                        int i;
+
+                        if (*format == '*')
+                                i = va_arg(*ap, int);
+                        else {
+                                uint64_t u;
+                                if (!parse_number8(format, &u, &format) || u > INT_MAX)
+                                        assert_not_reached();
+                                format--; /* Point it back to the last digit. */
+                                i = u;
+                        }
+
+                        if (have_field_width)
+                                precision = i;
+                        else
+                                field_width = i;
+                        break;
+                }
+
+                case '%':
+                        *ret_str16 = u"%";
+                        return format;
+
+                case 'c':
+                        /* char/char16_t/wchat_t get promoted to int/wint_t. */
+                        assert(field_length == sizeof(int) || long_arg);
+                        assert_cc(sizeof(int) >= sizeof(wint_t));
+
+                        tmp[0] = va_arg(*ap, int);
+                        tmp[1] = '\0';
+                        *ret_str16 = tmp;
+                        *ret_field_width = field_width;
+                        return format;
+
+                case 's':
+                        assert(field_length == sizeof(int) || long_arg);
+
+                        if (long_arg)
+                                *ret_wstr = va_arg(*ap, const wchar_t *);
+                        else
+                                *ret_str8 = va_arg(*ap, const char *);
+
+                        *ret_precision = precision;
+                        *ret_field_width = field_width;
+                        return format;
+
+                case 'm':
+                        if (status <= ELEMENTSOF(warn_table) - 1) {
+                                *ret_str8 = warn_table[status];
+                                return format;
+                        }
+
+                        if (status >= EFI_ERROR_MASK &&
+                            status <= ((ELEMENTSOF(err_table) - 1) | EFI_ERROR_MASK)) {
+                                *ret_str8 = err_table[NOERR(status)];
+                                return format;
+                        }
```

Comment:
i'd really add a helper call for this that takes the status as input and returns a const char * if it finds it in either of the two tables